### DISCUSSION TITLE
determining the magnetization parameter of a cylinder

### DISCUSSION THREAD

**Discussion by azadbolour** [magpylib/magpylib#675] [Q&A] [answered]
I am trying to use magpylib to calculate the magnetic field produced by a vertical stack of axially-magnetized cylindrical magnets, for example, the magnets listed at: 

  https://www.kjmagnetics.com/products.asp?cat=1&pg=3

My understanding is that the stack is modeled as a collection, and each cylinder is modeled by using the API function:

magpy.magnet.Cylinder(magnetization, dimension, position, orientation, style)

Since the stack is vertical and magnetization is axial, I assume that the magnetization parameter for my cylinders would look like: 

  (0, 0, z-magnetization)

But I am not sure how to determine the "z-magnetization" needed by magpylib for a cylinder magnet that I purchase on the internet, for example, a magnet of the following specifications (copied from the kjmagnetics site)?

  Dimensions: 1/8" dia. x 7/8" thick
  Tolerances: ±0.004" x ±0.004"
  Material: NdFeB, Grade N42
  Plating/Coating: Ni-Cu-Ni (Nickel)
  Magnetization Direction: Axial (Poles on Flat Ends)
  Weight: 0.0466 oz. (1.32 g)
  Pull Force, Case 1: 0.88 lbs
  Pull Force, Case 2: 0.88 lbs
  Surface Field: 6583 Gauss
  Max Operating Temp: 176ºF (80ºC)
  Brmax: 13,200 Gauss
  BHmax: 42 MGOe

How can I use the above information to calculate the magnetization required by the Cylinder function of magpylib? 

On another note, I am not sure whether I need to use the orientation and style parameters of the Cylinder function for my magnets, and if so, what their values should be. For now I am using their defaults.

Many thanks.

Azad

**Comment by OrtnerMichael** [comment, marked as answer]
Hi @azadbolour 

The N42 material has a Remanence Br (they call it residual flux density) of Br = 13200G = 1320 mT, which is the amplitude of the Magpylib magnetization vector. Your diameter is 1/8" = 3.175mm. The height is 7/8"=22.225mm. The following code creates a Cylinder magnet with respect to those specifications:

```python
import magpylib as magpy

cyl = magpy.magnet.Cylinder(
    magnetization=(0,0,1320),
    dimension=(3.175, 22.225),
)

cyl.show(backend='pyvista')
```

![image](https://github.com/magpylib/magpylib/assets/47384012/15ff2505-7917-4c3d-ae49-45e8376326f0)

If you have a stack of such Cylinders you simply copy this object, move it to the right position and add it to a `Collection`. For example like this:

```python
import magpylib as magpy

cyl = magpy.magnet.Cylinder(
    magnetization=(0,0,1320),
    dimension=(3.175, 22.225),
)

stack = magpy.Collection(cyl)
for i in range(1,4):
    stack.add(cyl.copy(position=(0,0,i*22.225)))

stack.show(backend='pyvista')
```

![image](https://github.com/magpylib/magpylib/assets/47384012/b5fbc0ca-fd80-4e7f-82f0-a66a7d74dbe5)

This stack you can now use as a single large Magnet. Please be aware that two magnets stacked onto each other like this create the same field as a single magnet with the same magnetization vector and twice the length.

In reality one must also be aware that the datasheet values are rarely correct. Often you will find that Br is up to 20% lower than what is claimed in the datasheets.

Does this answer your question?

**Comment by azadbolour** [comment]
Thank you so much for your detailed response to my question, including the
code samples. I really appreciate your clear and prompt clarification of my
use case.

Azad

**Comment by OrtnerMichael** [comment]
In reality the magnetization is a bit lower than the datasheet Br as a result of self-induced "demagnetization". Typically 5-15 %, depending on the material and the geometry of your magnet. You can account for this with the "demagnetization factor".

Here is a link to a tool where you can compute demag factors for cuboids.

http://www.magpar.net/static/magpar/doc/html/demagcalc.html

In addition, there is a contribution from inhomogeneous magnetizations resulting from demagnetization, but this is typically less than 1% for materials with \mu_r <1.2

You can find some computations in the Appendix A3 of this open-access paper:

https://www.mdpi.com/1424-8220/20/23/6873


excert from the paper:

![image](https://github.com/magpylib/magpylib/assets/47384012/99deda82-2095-464e-ad64-dfa10d68854f)

**Comment by azadbolour** [comment]
Thanks for the heads up. My magnets are cylindrical. I am not sure if the
referenced mdpi paper's results extend to cylinders. But unfortunately the
paper is too technical for me. I guess I will just have to be aware of the
inherent demagnetization error in analyzing my experiments.

Azad

**Reply by OrtnerMichael** [reply]
If your cylinders are long then the demagnetization effect is low. You could compute the demag factor with help of the link I provided above for a similarly long Cuboid, this would give you at least an idea of the order of magnitude involved.

If you compare to magnetic field sensor reads, be aware that Hall cells average over relatively large areas (typically 100 um) which will result in a "smaller peaks" as well.

Also: The reason I posted this above just now is because in the last 3 weeks we had three requests similar to yours - I will write a nice doku example with experimental data, but until then I try to collect all information here.

**Comment by luizenger** [comment]
@azadbolour I will start writing a doc on how to better make comparisons between Magpylib simulations and experiments with real magnets, ex: which value to use in Magpylib magnetization/polarization coming from a magnet datasheet

**Comment by azadbolour** [comment]
That sounds great. Looking forward to reading your document.

Azad

**Comment by luizenger** [comment]
**Please keep in mind that the following document is in a crude state as it is a work in progress. This is a "living document" that will be constantly updated and will be integrated in Magpylib's webpage.**

# Understanding Magnets
Whenever you wish to compare Magpylib simulations with experimental data obtained using a real permanent magnet, you might wonder how to properly set up the Magpylib magnet object to better reflect the physical permanent magnet in question.
First of all, it is important to get an understanding on how to read a permanent magnet datasheet, and then which value shall be used as the Magpylib's magnet magnetization/polarization attribute.

## Hysteresis Loop
If you've worked with magnetism, chances are very high that you have seen a hysteresis loop. This curve presents the behavior of the permanent magnet's *magnetic polarization* **J** (or sometimes the *average magnetization* **M**, related through µ<sub>0</sub>) and the resulting *magnetic flux density* **B** according to an external applied *magnetic field strength* **H**. While **B** and **H** exist everywhere, **J** only exists inside the material. The relation between these quantities is **B = µ<sub>0</sub>H + J**. The curve that describes the relationship between **J** and **H** is known as the *intrisic curve*, whereas the relationship between **B** and **H** is known as the *normal curve*, also called B-H curve. The normal curve represents the resulting **B**, considering contributions both by the magnet's magnetic polarization **J** and the external applied field **H** (wording can get confusing in Magnetism, but I am confident you can follow).
![hysteresis](https://github.com/magpylib/magpylib/assets/28402185/30d11ecf-d016-49d5-8acc-90cd25c172f8)
With the increase of **H** starting in the first quadrant, the material's polarization will increase up to its maximum possible value, the saturation **J<sub>S</sub>**. Higher values of **H** will not affect **J**, while the resulting **B** will keep increasing. This is reflected in the flat line of the intrinsic curve and the continuing higher normal curve. As the external field drops to zero, the material will retain its magnetization saturation while resulting **B** decreases. The magnetic flux density value when **H = 0** is known as the *remanence field* **B<sub>r</sub>**, and its only contribution is **J<sub>S</sub>**. As the external field goes negative (that is, its amplitude increases with a direction opposite to the magnetic polarization), we move towards the second quadrant of the hysteresis loop. With negative **H** and unchanged **J**, **B** reduces and the normal curve drops. At the *coercive field* **H<sub>c</sub>**, the net magnetic flux density in the material will be zero, but it still is magnetized! The material will retain its magnetization until it quickly drops to zero as **H** reaches a given value known as *intrinsic coercive field* **H<sub>ci</sub>**. This value represents how strongly a material can maintain its magnetization. As **H** gets stronger past **H<sub>ci</sub>**, polarization quickly aligns with the external field and the material once again reaches magnetization saturation, but now in the opposite direction from the previous state. If the external applied field goes towards positive values, we have a mirrored behavior and the hysteresis loop is closed.

## So what are we measuring?
When a magnetic field sensor is placed close to a permanent magnet, its output will reflect the magnetic flux density **B** acting on it. In other words, it will behave according to the normal curve of the hysteresis loop! If in your application the applied external field can be considered zero, it seems intuitive to just get the remanence **B<sub>r</sub>** value from the magnet's datasheet and use that as the polarization value in your Magpylib simulation. But when comparing the simulation results with experimental data, discrepancies arise. The measured **B** is smaller, and you already verified that other parameters such as the airgap are correct! So, what can it be?

## Datasheet of a permanent magnet
Let's imagine you are working with a sintered NdFeB permanent magnet from Bomatec (https://www.bomatec.com/en), as an example. You will easily find values of remanence, coercivity, operating temperature, density and others, usually presented by the magnet supplier in a table like this:
![image-3](https://github.com/magpylib/magpylib/assets/28402185/d6c5ac8a-526f-4d8f-9974-e7bfee85efb9)
After a quick glance, you would use the nominal **B<sub>r</sub>** value in tesla as the magnetic polarization of the Magpylib magnet. After all, we saw in the hysteresis loop that this coincides with **J** when the field **H** is zero. But inside the datasheet of the chosen magnet, there is an interesting figure:
![image-1](https://github.com/magpylib/magpylib/assets/28402185/6070af6a-4925-453b-968a-b20672d8b7b5)
This figure shows a focus on the second quadrant of the hysteresis loop, where only negative values for **H** are present. For two magnet temperatures, the intrinsic and normal curves are shown, being intersected with additional straigh lines. To obtain the correct magnetic polarization value **J** for the simulation, this figure has to be studied. 

### When the external applied field is not really external
Consider a permanent magnet surrounded by non-magnetic medium, such as a magnet in air. Outside of it, the only contribution to **B** measured by a sensor can be **H**, as **J = 0**. The source of this field is the magnetic polarization itself, and as it "leaks" outside the magnet it is also known as the *stray field*. But it also exists inside the magnet itself, in a direction opposite to that of the polarization! Because it acts as if trying to reduce the polarization of the magnet itself, another name for this field is the *demagnetizing field*. The figure below, obtained using Magpylib, (see the streamplot tutorial here[link]), shows the comparison between computed **H** and **B** for a rectangular magnet with homogeneous magnetization.
![FieldsComp2](https://github.com/magpylib/magpylib/assets/28402185/344f8d28-e64f-4bc7-aa39-bf410addef47)
Thus, with a positive magnetic polarization and negative **H**, datasheets focus on the second quadrant of the hysteresis loop.

### Getting the correct polarization
Then how do we obtain the correct value to be used in Magpylib simulation? For an homogeneous magnetized magnet, this self demagnetizing field depends only on the geometry of the material, through the *permeance coefficient*. The grey straight lines crossing the intrinsic and normal curves represent this coefficient. 
Using a cylindrical magnet with an axial magnetization as an example, the measures of interest are its length L and diameter D. Knowing the L/D ratio, we follow the related permeance coefficient line and verify the point where it intersects the normal curve. This point is known as the *working point* of the permanent magnet. The corresponding magnetic flux density  **B<sub>W</sub>** is what a sensor would measure if it were inside the magnet. But to get what is the actual magnetic polarization of the magnet, we follow a vertical line straight up from the working point and verify where the intrinsic curve gets intersected. The corresponding magnetic polarizazion **J<sub>W</sub>** is the value to be used when creating your Magpylib magnet.
![image-5](https://github.com/magpylib/magpylib/assets/28402185/32803049-29be-477a-ac7f-757efc980731)
Back to the datasheet: according to it, at room temperature the remanence **B<sub>r</sub>** is 1.22 T. If it were a cylinder magnet with L/D = 1.5, the polarization value in your Magpylib simulation has to be 1.2 T.
![bitmap](https://github.com/magpylib/magpylib/assets/28402185/109be2ca-97ef-4a72-8af7-e8aa5316cde2)
The difference in this case is small, but it can be much greater in many others! It depends on the grade, material, shape and quality of the magnet.

# TL;DR
With the datasheet of the permanent magnet in hands, look where the corresponding *permeance coefficient* line crosses the B-H curve. From there, trace a vertical line up and verify where the intrinsic curve gets intersected. That is the magnetic polarization value you should use in Magpylib simulations.

## Example with experimental data
**IN PROGRESS**

## Demagnetizing field origin

## Open magnetic circuits vs closed magnetic circuit

#### Exterior reference
G. Martinek, S. Ruoho and U. Wyss. (2021).*Magnetic Properties of Permanents Magnets & Measuring Techniques* [White paper]. Arnold Magnetic Technologies. https://www.arnoldmagnetics.com/blog/measuring-permanent-magnets-white-paper/

**Reply by luizenger** [reply]
@azadbolour Uploaded the first draft of the document, take a look and feel free to provide feedback!
